src/torchcodec/_core/CudaDeviceInterface.cpp

```diff
@@ -199,12 +199,121 @@ void CudaDeviceInterface::initializeContext(AVCodecContext* codecContext) {
+  // us.
+  if (avFrame->format != AV_PIX_FMT_CUDA) {
+    return nullptr;
+  }
+
+  TORCH_CHECK(
+      avFrame->hw_frames_ctx != nullptr,
+      "The AVFrame does not have a hw_frames_ctx. "
+      "That's unexpected, please report this to the TorchCodec repo.");
+
+  auto hwFramesCtx =
+      reinterpret_cast<AVHWFramesContext*>(avFrame->hw_frames_ctx->data);
+  AVPixelFormat actualFormat = hwFramesCtx->sw_format;
+
+  // NV12 conversion is implemented directly with NPP, no need for filters.
+  if (actualFormat == AV_PIX_FMT_NV12) {
+    return nullptr;
+  }
+
+  auto frameDims =
+      getHeightAndWidthFromOptionsOrAVFrame(videoStreamOptions, avFrame);
+  int height = frameDims.height;
+  int width = frameDims.width;
+
+  AVPixelFormat outputFormat;
+  std::stringstream filters;
+
+  unsigned version_int = avfilter_version();
+  if (version_int < AV_VERSION_INT(8, 0, 103)) {
+    // Color conversion support ('format=' option) was added to scale_cuda from
+    // n5.0. With the earlier version of ffmpeg we have no choice but use CPU
+    // filters. See:
+    // https://github.com/FFmpeg/FFmpeg/commit/62dc5df941f5e196164c151691e4274195523e95
+    outputFormat = AV_PIX_FMT_RGB24;
+
+    filters << "hwdownload,format=" << av_pix_fmt_desc_get(actualFormat)->name;
+    filters << ",scale=" << width << ":" << height;
+    filters << ":sws_flags=bilinear";
+  } else {
+    // Actual output color format will be set via filter options
+    outputFormat = AV_PIX_FMT_CUDA;
+
+    filters << "scale_cuda=" << width << ":" << height;
+    filters << ":format=nv12:interp_algo=bilinear";
+  }
+
+  return std::make_unique<FiltersContext>(
+      avFrame->width,
+      avFrame->height,
+      frameFormat,
+      avFrame->sample_aspect_ratio,
+      width,
+      height,
+      outputFormat,
+      filters.str(),
+      timeBase,
+      av_buffer_ref(avFrame->hw_frames_ctx));
+}
+
 void CudaDeviceInterface::convertAVFrameToFrameOutput(
     const VideoStreamOptions& videoStreamOptions,
     [[maybe_unused]] const AVRational& timeBase,
-    UniqueAVFrame& avFrame,
+    UniqueAVFrame& avInputFrame,
     FrameOutput& frameOutput,
     std::optional<torch::Tensor> preAllocatedOutputTensor) {
+  std::unique_ptr<FiltersContext> newFiltersContext =
+      initializeFiltersContext(videoStreamOptions, avInputFrame, timeBase);
+  UniqueAVFrame avFilteredFrame;
+  if (newFiltersContext) {
+    // We need to compare the current filter context with our previous filter
+    // context. If they are different, then we need to re-create a filter
+    // graph. We create a filter graph late so that we don't have to depend
+    // on the unreliable metadata in the header. And we sometimes re-create
+    // it because it's possible for frame resolution to change mid-stream.
+    // Finally, we want to reuse the filter graph as much as possible for
+    // performance reasons.
+    if (!filterGraph_ || filtersContext_ != newFiltersContext) {
```

Comment:
This is a copy-paste of the conditions from the CPU side, but `filtersContext_` and `newFiltersContext` are now pointers, not the actual objects. So `filtersContext_ != newFiltersContext` is comparing the _pointer values_, not the actual values in the context. We need `*filtersContext_ != *newFiltersContext`.